feat(Controller): add ability to handle binary like values

christianalfoni opened the pull request:
Allrighty... after long and good discussions on handling binary like things (not serializable stuff) I think I found a really good solution.

We talked about identifying these on the input or use a special `.keyName` convention... but we already check the keys if they are serializable, so only thing now is to handle it differently when they are not (The payload to signal itself still has a check on being serializable though, like it needs to be a plain object):

```js
someSignal({
  file: event.target.files[0]
})
```

There is so little code it is silly, the only thing it does when it is not serializable is that it attaches a `toJSON` method to the object :) 

In actual state tree it looks like:
```js
{
  file: FileObject
}
```

And when serialized it looks like:
```js
{
  files: '[NON_SERIALIZABLE]'
}
```

Feel free to suggest some other naming there... but now we do not even have to mention binary. We just allow you to pass in whatever on the keys. Some is serializable some is not. We just note to people that time travel debugging does not work with "non serializable" state :)

christianalfoni commented:
It might be an idea here actually to discard "non serializable" payload objects... not throwing errors. People have a tendency to not handle normal clicks etc. correctly. That is not where people go wrong... like "where is my event object?", it is that they are passing something in that they did not realise. 

So yeah... just ignoring those payloads I think would be good! Please comment :)

gaspard commented on packages/cerebral/src/utils.js:
```diff
@@ -107,3 +111,19 @@ export function debounce (func, wait) {
+    if (!isSerializable(currentPayload[key])) {
+      Object.defineProperty(currentPayload[key], 'toJSON', {
+        value () {
+          return '[NON_SERIALIZABLE]'
```
Maybe the value can be created only once.

christianalfoni replied:
It is only created once? :) It happens before signal runs... or do I misunderstand?

gaspard commented on packages/cerebral/src/Controller.js:
```diff
@@ -188,8 +172,8 @@ class Controller extends EventEmitter {
 
-    return function (payload) {
-      this.runSignal(path, signal, payload)
+    return function (payload, binaries) {
```
Not sure why this is here... :-)

christianalfoni replied:
Ops, merge issue, will fix!

gaspard commented:
For me ignoring events passed in signals is not so good. It's something annoying at first but it makes the code more readable:

```js
<button onClick={() => mySignal()} /> // Makes it obvious what is passed in sighal.
```

gaspard commented:
I like this solution most. It has many advantages:

- It is the least intrusive 
- it does not require any new API
- it works as expected to newcomers

christianalfoni commented:
@gaspard About wrong payload object, I was thinking we throw it away, BUT warn about it:

"You passed in an invalid payload object on signal"

This allows us to also log out the object that was invalid.

In devtools settings we have an option: `{warnInvalidPayload: true}`, which of course is true by default... basically, using warning and log object, rather than throwing error :)

Guria commented:
@christianalfoni please document it as well :)

christianalfoni commented:
Oh, of course :-) let me do another rebase!
On Wed, 30 Nov 2016 at 20:32, Aleksey Guryanov <notifications@github.com>
wrote:

> @christianalfoni <https://github.com/christianalfoni> please document it
> as well :)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cerebral/cerebral/pull/528#issuecomment-263971065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADxgwc3TnfWC74pihKkwHDp6PKF_GsZtks5rDc9CgaJpZM4LAZg8>
> .
>

christianalfoni commented:
Okay, lets see. I got some thinking down trying to calm down the little misses with stomach ache :)

We do want to warn when invalid values are put into the state tree. The current change gave you ability to just put whatever in, which is not really what we want. To allow both "special values" and still warn about wrong usage, these changes were made:

- Now Cerebral only checks serializability when devtools is active. Removed the `enforceSerializability` option, it is always on, but now it does not run in production

- If not sending in a valid base object as payload to signal you get a warning and the payload is ignored. This will avoid app just stopping, but rather educate. Typically people do not know they pass in the event object on clicks, now you just get a warning and nothing is added to payload. This also fixes issue where you pass in something serializable, but is not base object

- Added `allowedTypes` as option to devtools. By default **File**, **FileList** and **Blob** is there, but you can extend to whatever you want... allowing more special types. Please add more base types if you think it is relevant

- Any allowed types not having a `toJSON` ability gets one added based on the constructor name. To allow debugger to show it

- Updated docs

Okay, I think this has "all the things" :)

Guria commented on packages/docs/content/in-depth/03_state.en.md:
````diff
@@ -50,4 +50,13 @@ const conroller = Controller({
+date.setHours(8)
+```
+
+There is no way for Cerebral to keep track of this, and therefor is not supported. But do not worry about this. You can still use the Date object in your app, only that when you store it as state you use either a string or a number.
````
therefore

Guria commented on packages/docs/content/in-depth/03_state.en.md:
```diff
@@ -50,4 +50,13 @@ const conroller = Controller({
+
+There is no way for Cerebral to keep track of this, and therefor is not supported. But do not worry about this. You can still use the Date object in your app, only that when you store it as state you use either a string or a number.
+
+And that is basically all there is to state in Cerebral. If you know how to work with basic JavaScript types, you can work with state in Cerebral.
```
Hm. Aren't we have to use special api to deal with state?

christianalfoni replied:
Yeah, bad wording, changed it :)

Guria commented on packages/docs/content/in-depth/04_signals.en.md:
````diff
@@ -56,3 +56,5 @@ connect({
 )
 ```
+
+The payload passed to a component is typically the main value types of JavaScript. Object, Array, String, Number or Boolean. It is also possible to pass in some special value types, like files. For a full list of supported value types, check the state API documentation.
````
link to state API here?

christianalfoni replied:
Jup, done!

Guria commented on packages/docs/content/api/02_state.en.md:
```diff
@@ -19,6 +19,15 @@ const controller = Controller({
+
 ### Get state
 The only way to get state in your application is by connecting it to a component or grabbing it in an action.
 
```
Could you please also document state.compute here?

christianalfoni replied:
For sure :)

christianalfoni commented:
Okay, updated your comments @Guria and made some small improvements myself as well :)
